File: FirstAndLastIndexOfElement.java
class Solution {
    private int binarySearch(int[] nums,int target,int low,int high,boolean isLeft){
        while(low<=high){
            int mid = low+(high-low)/2;
            if(isLeft){
                if(nums[mid]==target){
                    if(mid==0||nums[mid]>nums[mid-1]){
                        return mid;
                    }else{  //moving towards left
                        high = mid-1;
                    }
                }else if(nums[mid]>target){
                   high = mid-1;
                }else{
                    low = mid+1;
                }
            }else{
                if(nums[mid]==target){
                    if(mid==0||nums[mid]<=nums[mid+1]){
                        return mid;
                    }else{//moving towards right
                        low = mid+1;
                    }
                }else if(nums[mid]<target){
                    low = low+1;
                }else{
                    high = high -1;
                }
            }
        }
        return -1;
    }
    public int[] searchRange(int[] nums,int target){
        if(nums == null||nums.length==0) return new int[]{-1,-1};
        if(target<nums[0]||target>nums[nums.length-1]) return new int[]{-1,-1};
        int left = binarySearch(nums,target,0,nums.length-1,true);
        int right = binarySearch(nums,target,left,nums.length-1,false);
        return new int[]{left,right};
    }
}

File: first_and_last_index_of_element.py
# Time Complexity : O(log(n^2)) because we are doing BS two time on same array
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : YES
# Any problem you faced while coding this : NO


from typing import List

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if nums == None or not nums: return [-1,-1]
        first = self.binarySearchFirst(nums, target)
        if first == -1: return [-1,-1]
        last = self.binarySearchLast(nums, target, first)
        return [first, last]

    def binarySearchFirst(self, nums, target):
        low, high = 0, len(nums) - 1
        while (low <= high):
            mid = (low + high) // 2
            if nums[mid] == target:
                # we'll check if mid the first index of target ??
                if mid == 0 or nums[mid] != nums[mid - 1]: return mid
                else: high = mid - 1
            elif nums[mid] > target:
                high = mid - 1
            else: low = mid + 1
        return -1 
    
    def binarySearchLast(self, nums, target, first):
        low, high = first, len(nums) - 1
        while (low <= high):
            mid = (low + high) // 2
            if nums[mid] == target:
                if mid == high or nums[mid] != nums[mid + 1]: return mid
                else: low = mid + 1
            elif nums[mid] > target:
                high = mid - 1
            else:
                low = mid + 1
        return -1